virt-v2v/cold/entrypoint.go

```diff
@@ -0,0 +1,274 @@
+package main
+
+import (
+	"context"
+	"encoding/xml"
+	"fmt"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	OVA     = "ova"
+	vSphere = "vSphere"
+	DIR     = "/var/tmp/v2v"
+	FS      = "/mnt/disks/disk[0-9]*"
+	Block   = "/dev/block[0-9]*"
+	VDDK    = "/opt/vmware-vix-disklib-distrib"
+)
+
+type Domain struct {
+	XMLName xml.Name `xml:"domain"`
+	Name    string   `xml:"name"`
+	OS      OS       `xml:"os"`
+}
+
+type OS struct {
+	Type   OSType `xml:"type"`
+	Loader Loader `xml:"loader"`
+	Nvram  Nvram  `xml:"nvram"`
+}
+
+type OSType struct {
+	Arch    string `xml:"arch,attr"`
+	Machine string `xml:"machine,attr"`
+	Content string `xml:",chardata"`
+}
+
+type Loader struct {
+	Readonly string `xml:"readonly,attr"`
+	Type     string `xml:"type,attr"`
+	Secure   string `xml:"secure,attr"`
+	Path     string `xml:",chardata"`
+}
+
+type Nvram struct {
+	Template string `xml:"template,attr"`
+}
+
+var (
+	xmlFilePath      string
+	requestProcessed = make(chan bool)
+)
+
+func main() {
+	virtV2vArgs := []string{"virt-v2v", "-v", "-x"}
+	source := os.Getenv("V2V_source")
+
+	requiredEnvVars := map[string][]string{
+		vSphere: {"V2V_libvirtURL", "V2V_secretKey", "V2V_vmName"},
+		OVA:     {"V2V_diskPath", "V2V_vmName"},
+	}
+
+	if envVars, ok := requiredEnvVars[source]; ok {
+		if !checkEnvVariablesSet(envVars...) {
+			fmt.Printf("Following environment variables need to be defined: %v\n", envVars)
+			os.Exit(1)
+		}
+	}
+
+	switch source {
+	case vSphere:
+		virtV2vArgs = append(virtV2vArgs, "-i", "libvirt", "-ic", os.Getenv("V2V_libvirtURL"))
+		pwFilePath := fmt.Sprintf("%s/vmware.pw", DIR)
+		if err := os.WriteFile(pwFilePath, []byte(os.Getenv("V2V_secretKey")), 0600); err != nil {
+			fmt.Printf("Failed to write to disk: %v\n", err)
+			os.Exit(1)
+		}
+		virtV2vArgs = append(virtV2vArgs, "-ip", pwFilePath)
+
+		if _, err := os.Stat(VDDK); err == nil {
+			virtV2vArgs = append(virtV2vArgs,
+				"-it", "vddk",
+				"-io", fmt.Sprintf("vddk-libdir=%s", VDDK),
+				"-io", fmt.Sprintf("vddk-thumbprint=%s", os.Getenv("V2V_thumbprint")),
+			)
+		}
+		virtV2vArgs = append(virtV2vArgs, "--", os.Getenv("V2V_vmName"))
+	case OVA:
+		virtV2vArgs = append(virtV2vArgs, "-i", "ova", os.Getenv("V2V_diskPath"))
+	}
+
+	fmt.Println("Preparing virt-v2v")
+	if err := os.MkdirAll(DIR, os.ModePerm); err != nil {
+		fmt.Println("Error creating directory  ", err)
+		os.Exit(1)
+	}
+	virtV2vArgs = append(virtV2vArgs, "-o", "local", "-os", DIR)
+
+	//Disks on filesystem storage.
+	if err := LinkDisks(FS, 15); err != nil {
+		os.Exit(1)
+	}
+	//Disks on block storage.
+	if err := LinkDisks(Block, 10); err != nil {
+		os.Exit(1)
+	}
+
+	if err := executeVirtV2v(virtV2vArgs); err != nil {
+		os.Exit(1)
+	}
+
+	if source == OVA {
+		var err error
+		xmlFilePath, err = waitForXMLFile(DIR, "xml", 5, 60)
```

Comment:
as discussed, if the XML is expected to be present when virt-v2v finishes there is no need to wait

Reply:
yeah, the OVF should be available when the conversion is done

Reply:
I've added it to avoid any possibility of condition race, but I think I just check if the command really ended rather than if the file was created.